[crypto] add instructions in K-ext spec v0.9.0

WuSiYu opened the pull request:
add instructions (and their builtins) in K-ext spec v0.9.0, except those in Zkb and Zkg which is waiting for ext-b's work to prevent conflict.

kito-cheng commented on gcc/common/config/riscv/riscv-common.c:
```diff
@@ -645,6 +651,19 @@ static const riscv_ext_flag_table_t riscv_ext_flag_table[] =
   {"d", &gcc_options::x_target_flags, MASK_DOUBLE_FLOAT},
   {"c", &gcc_options::x_target_flags, MASK_RVC},
+
+  {"k", &gcc_options::x_riscv_crypto_subext, MASK_ZKNE | MASK_ZKND | MASK_ZKNH | MASK_ZKG | MASK_ZKB | MASK_ZKR},
```
I would suggest you using `riscv_implied_info` to record the implied extension info like `zb*`, you could reference her e: https://github.com/riscv/riscv-gcc/blob/riscv-gcc-10.2.0-rvb/gcc/common/config/riscv/riscv-common.c#L63

kito-cheng commented on gcc/common/config/riscv/riscv-common.c:
```diff
@@ -645,6 +651,19 @@ static const riscv_ext_flag_table_t riscv_ext_flag_table[] =
+  {"zkg", &gcc_options::x_riscv_crypto_subext, MASK_ZKG},
+  {"zkb", &gcc_options::x_riscv_crypto_subext, MASK_ZKB},
+  {"zkr", &gcc_options::x_riscv_crypto_subext, MASK_ZKR},
+  {"zkn", &gcc_options::x_riscv_crypto_subext, MASK_ZKNE | MASK_ZKND | MASK_ZKNH | MASK_ZKG | MASK_ZKB},
```
Ditto.

kito-cheng commented on gcc/common/config/riscv/riscv-common.c:
```diff
@@ -645,6 +651,19 @@ static const riscv_ext_flag_table_t riscv_ext_flag_table[] =
+  {"zkne", &gcc_options::x_riscv_crypto_subext, MASK_ZKNE},
+  {"zknd", &gcc_options::x_riscv_crypto_subext, MASK_ZKND},
+  {"zknh", &gcc_options::x_riscv_crypto_subext, MASK_ZKNH},
+  {"zks", &gcc_options::x_riscv_crypto_subext, MASK_ZKSED | MASK_ZKSH | MASK_ZKG | MASK_ZKB},
```
Ditto.

kito-cheng reviewed (commented):
I would suggest you add at least one test case for each sub-extension of intrinsic functions, to make sure the code gen is expected.

WuSiYu commented on gcc/config/riscv/crypto.md:
```diff
@@ -0,0 +1,302 @@
+;; Zkr - Entropy Source
+
+(define_insn "riscv_pollentropy_<mode>"
+  [(unspec_volatile [(match_operand:X 0 "register_operand" "=r")] UNSPEC_CRYPTO)]
```
I tried to changed this to `(set (match_operand:X 0 "register_operand" "=r") (unspec:X UNSPEC_XXX))`, but it seem that it can't pass the compile, the "unspec" wants a vector after it.

kito-cheng replied:
Use a dummy `const_int 0` for that, like  `(set (match_operand:X 0 "register_operand" "=r") (unspec:X [(const_int 0)] UNSPEC_XXX))`, this trick are used in several different back-end :)

You can grep by following command to see:
```
grep "\[(const_int 0)\]" * -R |grep unspec
```

jim-wilson commented on gcc/config/riscv/crypto.md:
```diff
@@ -0,0 +1,302 @@
+;; Machine description for K extension.
+;; Copyright (C) 2011-2021 Free Software Foundation, Inc.
```
This is a new file.  It only needs to mention the current year.

jim-wilson commented on gcc/config/riscv/crypto.md:
```diff
@@ -0,0 +1,302 @@
+
+(define_insn "riscv_aes32dsi"
+  [(set (match_operand:SI 0 "register_operand" "=r")
+    (unspec:SI
```
If you are an emacs user, you can do M-x lisp-mode and then tab will indent the RTL properly.  The open parenthesis here should align with the (match_operand on the line above.

jim-wilson reviewed (commented):
Just some minor comments to add to a lot of good comments from Kito.

WuSiYu commented:
Thanks for your comments and guidance! I pushed a new commit to fix these. (And to make the `riscv_implied_info` works, I also cherry-pick a patch from `riscv-gcc-10.2.0-bump` branch.)
I think I fixed most of them, but there are still some questions that may need your guidance.

@kito-cheng For the UNSPEC_* part, I split them into the things like `UNSPEC_AES_DS`, `UNSPEC_AES_DS` ... ,I am not sure if this is too sketchy or too detailed? please tell me if I did it wrong.

@jim-wilson Unfortunately I am not familiar with emacs, but I found a tool that use emacs to format lisp code, I use it on `crypto.md` and hope it worked fine for the .md.

kito-cheng commented:
This patch is LGTM now.
One minor suggestion is you could using `define_int_iterator` + `define_int_attr` to further simplify RTL pattern, but it's not must to do.
